### PR TITLE
[WIP] Adding `ip_discovery` and `network_type` for elasticache replication group

### LINKED ISSUE #28131: [aws_elasticache_replication_group]: add support for `network_type` and `ip_discovery` arguments
### Description

It would be great to add this to the clustered mode just like you have it for `elasticache_cluster`.

### Affected Resource(s) and/or Data Source(s)

_No response_

### Potential Terraform Configuration

_No response_

### References

_No response_

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by jms200**
### Description

Adds the ip_discovery parameter to elasticache replication group that is now supported on Redis 6.2 and Memcached engine version 1.6.6 on all instances built on the [Nitro system](https://aws.amazon.com/ec2/nitro/).

### Relations

Closes #28131 

### References

API Documentation: https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_ReplicationGroup.html

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

### CODE REVIEW

**Comment by jms200** [comment]
This is still a work in progress and need to get the tests written for the replication group.

File: internal/service/elasticache/replication_group.go [hunk 1]
```diff
@@ -461,6 +474,14 @@ func resourceReplicationGroupCreate(d *schema.ResourceData, meta interface{}) er
 		params.CacheParameterGroupName = aws.String(v.(string))
 	}
 
+	if v, ok := d.GetOk("ip_discovery"); ok {
+		req.IpDiscovery = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("network_type"); ok {
+		req.NetworkType = aws.String(v.(string))
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group.go:482]: use `params` instead of `req`

Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group.go:482]: use `params` instead of `req`

```suggestion
		params.NetworkType = aws.String(v.(string))
```

File: internal/service/elasticache/replication_group.go [hunk 2]
```diff
@@ -461,6 +474,14 @@ func resourceReplicationGroupCreate(d *schema.ResourceData, meta interface{}) er
 		params.CacheParameterGroupName = aws.String(v.(string))
 	}
 
+	if v, ok := d.GetOk("ip_discovery"); ok {
+		req.IpDiscovery = aws.String(v.(string))
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group.go:478]: use `params` instead of `req`

```suggestion
		params.IpDiscovery = aws.String(v.(string))
```

File: internal/service/elasticache/replication_group.go [hunk 3]
```diff
@@ -677,6 +698,9 @@ func resourceReplicationGroupRead(d *schema.ResourceData, meta interface{}) erro
 	d.Set("arn", rgp.ARN)
 	d.Set("data_tiering_enabled", aws.StringValue(rgp.DataTiering) == elasticache.DataTieringStatusEnabled)
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group.go:700]: you can remove the whiteline above and below the newly added attributes.

File: internal/service/elasticache/replication_group.go [hunk 4]
```diff
@@ -816,6 +840,16 @@ func resourceReplicationGroupUpdate(d *schema.ResourceData, meta interface{}) er
 		requestUpdate = true
 	}
 
+	if d.HasChange("ip_discovery") {
+		params.IpDiscovery = aws.String(d.Get("ip_discovery").(string))
+		requestUpdate = true
+	}
+
+	if d.HasChange("network_type") {
+		req.IpDiscovery = aws.String(d.Get("network_type").(string))
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group.go:849]: use `params` instead of `req`

```suggestion
		req.IpDiscovery = aws.String(d.Get("network_type").(string))
```

File: internal/service/elasticache/replication_group_test.go [hunk 5]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
 	})
 }
 
+func TestAccElastiCacheReplicationGroup_network_type_ipv6(t *testing.T) {
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:69]: You can rename the function to `TestAccElastiCacheReplicationGroup_networkTypeIPV6`

File: internal/service/elasticache/replication_group_test.go [hunk 6]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
...
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, elasticache.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckReplicationGroupDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccReplicationGroupConfig_basic(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:86]: Get rid of some of the less interesting attributes down here, just focus on what's mandatory to check + what the test covers.

File: internal/service/elasticache/replication_group_test.go [hunk 7]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
...
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"apply_immediately"}, //not in the API
+			},
+		},
+	})
+}
+
+func TestAccElastiCacheReplicationGroup_network_type_dual_stack(t *testing.T) {
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:117]: You can rename the function to `TestAccElastiCacheReplicationGroup_networkTypeDualStack`

File: internal/service/elasticache/replication_group_test.go [hunk 8]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
...
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, elasticache.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckReplicationGroupDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccReplicationGroupConfig_basic(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:134]: Get rid of some of the less interesting attributes down here, just focus on what's mandatory to check + what the test covers.

File: website/docs/d/elasticache_replication_group.html.markdown [hunk 9]
```diff
@@ -32,6 +32,8 @@ In addition to all arguments above, the following attributes are exported:
 * `arn` - ARN of the created ElastiCache Replication Group.
 * `auth_token_enabled` - Whether an AuthToken (password) is enabled.
 * `automatic_failover_enabled` - A flag whether a read-only replica will be automatically promoted to read/write primary if the existing primary fails.
+* `ip_discovery` - The IP version advertised in the discovery protocol.
```
Comment (bschaatsbergen) [inline website/docs/d/elasticache_replication_group.html.markdown:35]: Can you remove the changes you made in the data source docs, you haven't actually updated the data sources. Lets do that through a different PR.

**Review by bschaatsbergen — CHANGES_REQUESTED** [review]

File: internal/service/elasticache/replication_group_test.go [hunk 10]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
...
+	resourceName := "aws_elasticache_replication_group.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, elasticache.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckReplicationGroupDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccReplicationGroupConfig_basic(rName),
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:133]: Be sure to create your own config per test, and ensure that you set the respective attributes that you want to test too.

File: internal/service/elasticache/replication_group.go (hunk 1 above)
Comment (jms200) [reply internal/service/elasticache/replication_group.go]: Completed

File: internal/service/elasticache/replication_group.go (hunk 2 above)
Comment (jms200) [reply internal/service/elasticache/replication_group.go]: Completed

File: internal/service/elasticache/replication_group.go (hunk 3 above)
Comment (jms200) [reply internal/service/elasticache/replication_group.go:700]: Completed

File: internal/service/elasticache/replication_group_test.go (hunk 5 above)
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go:69]: Completed

File: internal/service/elasticache/replication_group_test.go (hunk 6 above)
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go:86]: Completed

File: internal/service/elasticache/replication_group_test.go (hunk 8 above)
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go:134]: Completed

File: website/docs/d/elasticache_replication_group.html.markdown (hunk 9 above)
Comment (jms200) [reply website/docs/d/elasticache_replication_group.html.markdown:35]: Completed

File: internal/service/elasticache/replication_group_test.go [hunk 11]
```diff
@@ -66,6 +66,102 @@ func TestAccElastiCacheReplicationGroup_basic(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go]: Completed

File: internal/service/elasticache/replication_group_test.go (hunk 10 above)
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go]: Completed

File: internal/service/elasticache/replication_group_test.go [hunk 12]
```diff
@@ -3796,6 +3876,29 @@ data "aws_elasticache_replication_group" "test" {
 `, rName, enableClusterMode, slowLogDeliveryEnabled, slowDeliveryDestination, slowDeliveryFormat, engineLogDeliveryEnabled, engineDeliveryDestination, engineLogDeliveryFormat)
 }
 
+func testAccReplicationGroupConfig_networkTypeIPv6(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_elasticache_replication_group" "test" {
+  replication_group_id          = %[1]q
+  replication_group_description = "test description"
+  node_type                     = "cache.t3.small"
+  ip_discovery					= "ipv6"
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:3885]: can you fmt this

File: internal/service/elasticache/replication_group_test.go [hunk 13]
```diff
@@ -3796,6 +3876,29 @@ data "aws_elasticache_replication_group" "test" {
...
+}
+`, rName)
+}
+
+func testAccReplicationGroupConfig_networkTypeDualStack(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_elasticache_replication_group" "test" {
+  replication_group_id          = %[1]q
+  replication_group_description = "test description"
+  node_type                     = "cache.t3.small"
```
Comment (bschaatsbergen) [inline internal/service/elasticache/replication_group_test.go:3896]: can you fmt this

File: website/docs/r/elasticache_replication_group.html.markdown [hunk 14]
```diff
@@ -190,10 +190,12 @@ The following arguments are optional:
...
 * `maintenance_window` – (Optional) Specifies the weekly time range for when maintenance on the cache cluster is performed. The format is `ddd:hh24:mi-ddd:hh24:mi` (24H Clock UTC). The minimum maintenance window is a 60 minute period. Example: `sun:05:00-sun:09:00`
 * `multi_az_enabled` - (Optional) Specifies whether to enable Multi-AZ Support for the replication group. If `true`, `automatic_failover_enabled` must also be enabled. Defaults to `false`.
+* `network_type` - (Optional) The IP versions for cache cluster connections. IPv6 is supported with Redis engine `6.2` onword or Memcached version `1.6.6` for all [Nitro system](https://aws.amazon.com/ec2/nitro/) instances. Valid values are `ipv4`, `ipv6` or `dual_stack`.
```
Comment (bschaatsbergen) [inline website/docs/r/elasticache_replication_group.html.markdown:198]: I would probably remove the second sentence, this is an API limitation but might be fixed in the future (who knows). In that case our docs would be misleading.

File: .changelog/28306.txt [hunk 15]
```diff
@@ -0,0 +1,3 @@
+```release-note:enhancement
+resource/aws_elasticache_replication_group: Add `ip_discovery` and `network_type` arguments in support of [IPv6 clusters](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/network-type.html)
```
Comment (bschaatsbergen) [inline .changelog/28306.txt:2]: No need for the href here to `IPv6 clusters`

**Review by bschaatsbergen — CHANGES_REQUESTED** [review]

File: internal/service/elasticache/replication_group_test.go (hunk 13 above)
Comment (jms200) [reply internal/service/elasticache/replication_group_test.go:3896]: Formatted

File: website/docs/r/elasticache_replication_group.html.markdown (hunk 14 above)
Comment (jms200) [reply website/docs/r/elasticache_replication_group.html.markdown:198]: Completed

File: .changelog/28306.txt (hunk 15 above)
Comment (jms200) [reply .changelog/28306.txt:2]: Removed href

**Comment by jms200** [comment]
Currently working through issues with getting acceptance testing to complete.

**Comment by kedare** [comment]
Hello.

Is there any news on this ? Any help needed ?

Thanks.

**Comment by kbergha** [comment]
Hi. Any hope getting this updated for 5.x (if needed) and merged?

**Comment by larsboldt** [comment]
We need this to properly support IPv6 which is required by law in Norway - please prioritize merging.

**Comment by bschaatsbergen** [comment]
I'll take a stab at it @larsboldt

**Comment by bschaatsbergen** [comment]
I'm moving the fixes into a new branch and PR as I have no write access on the fork.
